Add CryptoKit implementations for HMAC/HKDF in WebKit

nmahendru opened the pull request:
#### be5cbef50e65beef09ae6956defb911329c8cb29
<pre>
Add CryptoKit implementations for HMAC/HKDF in WebKit
<a href="https://rdar.apple.com/126689803">rdar://126689803</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=272879">https://bugs.webkit.org/show_bug.cgi?id=272879</a>

Reviewed by Alex Christensen.

Add HMAC/HKDF implementations using CryptoKit. Still disabled by
default.

Also:
- disabled some sha224 tests. It&apos;s not part of the webcrypto standard
  anymore.(since 2017)
- Renamed some Swift classes to not use acronyms.

* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::makePinAuth):
(fido::pin::SetPinRequest::tryCreate):
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift.h:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(HMAC.sign(_:data:hashFunction:)):
(HMAC.verify(_:key:data:hashFunction:)):
(HKDF.deriveBits(_:salt:info:outputByteCount:hashFunction:)):
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(CryptoKit.isValidAuthenticationCode(_:data:key:)):
(CryptoKit.deriveKey(_:salt:info:outputByteCount:)):
* Source/WebCore/PAL/pal/PALSwiftUtils.h: Copied from Source/WebCore/PAL/pal/PALSwift.h.
(WebCore::toCKHashFunction):
(WebCore::isValidHashParameter):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.cpp:
(WebCore::CryptoAlgorithmHKDF::deriveBits):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp:
(WebCore::CryptoAlgorithmHMAC::sign):
(WebCore::CryptoAlgorithmHMAC::verify):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
(WebCore::toCKHashFunction): Deleted.
(WebCore::isValidHashParameter): Deleted.
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHKDFMac.cpp:
(WebCore::platformDeriveBitsCC):
(WebCore::platformDeriveBitsCryptoKit):
(WebCore::CryptoAlgorithmHKDF::platformDeriveBits):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:
(WebCore::platformSignCryptoKit):
(WebCore::platformVerifyCryptoKit):
(WebCore::platformSignCC):
(WebCore::platformVerifyCC):
(WebCore::CryptoAlgorithmHMAC::platformSign):
(WebCore::CryptoAlgorithmHMAC::platformVerify):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmHKDFGCrypt.cpp:
(WebCore::CryptoAlgorithmHKDF::platformDeriveBits):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmHMACGCrypt.cpp:
(WebCore::CryptoAlgorithmHMAC::platformSign):
(WebCore::CryptoAlgorithmHMAC::platformVerify):
* Source/WebCore/crypto/openssl/CryptoAlgorithmHKDFOpenSSL.cpp:
(WebCore::CryptoAlgorithmHKDF::platformDeriveBits):
* Source/WebCore/crypto/openssl/CryptoAlgorithmHMACOpenSSL.cpp:
(WebCore::CryptoAlgorithmHMAC::platformSign):
(WebCore::CryptoAlgorithmHMAC::platformVerify):

Canonical link: <a href="https://commits.webkit.org/277888@main">https://commits.webkit.org/277888@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d153ccab368c40cf9a984677475f604ae48eb7be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28047 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51801 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51521 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44900 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51139 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39923 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49416 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25698 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42112 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21024 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23161 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43292 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6889 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43782 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53432 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20154 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47227 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25148 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42321 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46174 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25956 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6978 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nmahendru commented on LayoutTests/TestExpectations:
```diff
@@ -2677,6 +2677,8 @@ crypto/subtle/ecdh-import-pkcs8-key-p384-validate-ecprivatekey-parameters-public
 crypto/subtle/ecdsa-import-pkcs8-key-p256-validate-ecprivatekey-parameters-publickey.html [ Skip ]
 crypto/subtle/ecdsa-import-pkcs8-key-p384-validate-ecprivatekey-parameters-publickey.html [ Skip ]
 crypto/subtle/ecdh-import-spki-key-ecdh-identifier.html [ Skip ]
+crypto/subtle/hmac-import-key-verify-sha224.html [ Skip ]
```
This should have been skipped long ago. sha-224 is no long in the webstandard since 2017. CryptoKit does not support SHA-224 so I am just pre-emptively disabling this so that it does not fail when we enable CryptoKit by default.

johnwilander replied:
Shouldn't we delete the test case all together then?

johnwilander commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -235,7 +235,8 @@ extension Curve25519.Signing.PrivateKey {
         }
-        return try self.signature(for: Data.temporaryDataFromSpan(spanNoCopy: span)).copyToVectorUInt8()
+        return try self.signature(for: Data.temporaryDataFromSpan(spanNoCopy: span))
+            .copyToVectorUInt8()
```
Was this line feed intended?

nmahendru replied:
https://github.com/WebKit/WebKit/pull/27474#discussion_r1576409180

johnwilander commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -250,8 +251,11 @@ extension Curve25519.Signing.PublicKey {
-        return self.isValidSignature(Data.temporaryDataFromSpan(spanNoCopy: signature), for: Data.temporaryDataFromSpan(spanNoCopy: data))
+        return self.isValidSignature(
+            Data.temporaryDataFromSpan(spanNoCopy: signature),
+            for: Data.temporaryDataFromSpan(spanNoCopy: data))
```
I'd agree that the code looks better this way but we often refrain from making cosmetic changes to actual code since the revision history gets obscured by it. With this change, it'll look at a glance like you wrote this code in your patch but it is older.

nmahendru replied:
https://github.com/WebKit/WebKit/pull/27474#discussion_r1576409180

johnwilander commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -265,9 +269,45 @@ extension Curve25519.KeyAgreement.PrivateKey {
         }
-        let pub =  try Curve25519.KeyAgreement.PublicKey(rawRepresentation: Data.temporaryDataFromSpan(spanNoCopy: pubSpan))
+        let pub = try Curve25519.KeyAgreement.PublicKey(
+            rawRepresentation: Data.temporaryDataFromSpan(spanNoCopy: pubSpan))
```
Ditto.

nmahendru replied:
All the cosmetic changes in swift files are a result of `Tools/Script/swift/format <file>`. 
I am just letting the tool decide rather than me deciding what is right. I might have missed running it in the previous commit.

johnwilander commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
```diff
@@ -95,6 +95,7 @@ ExceptionOr<Vector<uint8_t>> CryptoAlgorithmAESGCM::platformEncrypt(const Crypto
 
 ExceptionOr<Vector<uint8_t>> CryptoAlgorithmAESGCM::platformDecrypt(const CryptoAlgorithmAesGcmParams& parameters, const CryptoKeyAES& key, const Vector<uint8_t>& cipherText)
 {
+    // FIXME: One this is resolved then add decrypt with CryptoKit. rdar://92701544
```
Should say once, not one.

johnwilander replied:
I would also reverse it to "Add decryption with CryptoKit once this is resolved."

nmahendru replied:
done.

johnwilander commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
```diff
@@ -31,37 +31,13 @@
 namespace WebCore {
 #if HAVE(SWIFT_CPP_INTEROP)
 
-static PAL::HashFunction toCKHashFunction(CryptoAlgorithmIdentifier hash)
```
I guess this was moved, huh?

nmahendru replied:
This was moved to a header to avoid de-duplication as now it is needed at 3 different places.

johnwilander commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:
```diff
@@ -71,4 +89,25 @@ ExceptionOr<bool> CryptoAlgorithmHMAC::platformVerify(const CryptoKeyHMAC& key,
+    if (useCryptoKit == UseCryptoKit::Yes)
+        return platformSignCryptoKit(key, data);
+#else
+    UNUSED_PARAM(useCryptoKit);
```
This looks like an opportunity to log. Or are there valid cases where we'd get `UseCryptoKit::Yes` but `HAVE(SWIFT_CPP_INTEROP)` is false?

johnwilander replied:
Could even assert if it's really not supported.

nmahendru replied:
There are cases for `UseCryptoKit::No` but `HAVE(SWIFT_CPP_INTEROP)` yes.
Look at `CryptoKeyEC.h`. It changes the definition of platform key container based on `HAVE(SWIFT_CPP_INTEROP)` and `UseCryptoKit` can have both a `Yes` or a `No`

Some platforms that we run EWS on don't support interop and some platforms do and on both platforms in EWS UseCryptoKit == No. UseCryptoKit == Yes is only for my testing on my machine until we are ready to flip the flag.

johnwilander commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:
```diff
@@ -71,4 +89,25 @@ ExceptionOr<bool> CryptoAlgorithmHMAC::platformVerify(const CryptoKeyHMAC& key,
+    if (useCryptoKit == UseCryptoKit::Yes)
+        return platformVerifyCryptoKit(key, signature, data);
+#else
+    UNUSED_PARAM(useCryptoKit);
```
Ditto.

nmahendru replied:
https://github.com/WebKit/WebKit/pull/27474#discussion_r1576414773

achristensen07 commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -464,60 +464,170 @@ public enum EdKeyAgreementAlgorithm {
+        algo: EdSigningAlgorithm, key: SpanConstUInt8, signature: SpanConstUInt8,
+        data: SpanConstUInt8
+    ) -> EdReturnValue {
+        var rv = EdReturnValue()
```
It would be good to also rename "rv" to something that's not an abbreviation, like "result"

nmahendru committed:
Add CryptoKit implementations for HMAC/HKDF in WebKit
rdar://126689803
https://bugs.webkit.org/show_bug.cgi?id=272879

Reviewed by Alex Christensen.

Add HMAC/HKDF implementations using CryptoKit. Still disabled by
default.

Also:
- disabled some sha224 tests. It's not part of the webcrypto standard
  anymore.(since 2017)
- Renamed some Swift classes to not use acronyms.

* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::makePinAuth):
(fido::pin::SetPinRequest::tryCreate):
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift.h:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(HMAC.sign(_:data:hashFunction:)):
(HMAC.verify(_:key:data:hashFunction:)):
(HKDF.deriveBits(_:salt:info:outputByteCount:hashFunction:)):
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(CryptoKit.isValidAuthenticationCode(_:data:key:)):
(CryptoKit.deriveKey(_:salt:info:outputByteCount:)):
* Source/WebCore/PAL/pal/PALSwiftUtils.h: Copied from Source/WebCore/PAL/pal/PALSwift.h.
(WebCore::toCKHashFunction):
(WebCore::isValidHashParameter):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.cpp:
(WebCore::CryptoAlgorithmHKDF::deriveBits):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp:
(WebCore::CryptoAlgorithmHMAC::sign):
(WebCore::CryptoAlgorithmHMAC::verify):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
(WebCore::toCKHashFunction): Deleted.
(WebCore::isValidHashParameter): Deleted.
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHKDFMac.cpp:
(WebCore::platformDeriveBitsCC):
(WebCore::platformDeriveBitsCryptoKit):
(WebCore::CryptoAlgorithmHKDF::platformDeriveBits):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:
(WebCore::platformSignCryptoKit):
(WebCore::platformVerifyCryptoKit):
(WebCore::platformSignCC):
(WebCore::platformVerifyCC):
(WebCore::CryptoAlgorithmHMAC::platformSign):
(WebCore::CryptoAlgorithmHMAC::platformVerify):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmHKDFGCrypt.cpp:
(WebCore::CryptoAlgorithmHKDF::platformDeriveBits):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmHMACGCrypt.cpp:
(WebCore::CryptoAlgorithmHMAC::platformSign):
(WebCore::CryptoAlgorithmHMAC::platformVerify):
* Source/WebCore/crypto/openssl/CryptoAlgorithmHKDFOpenSSL.cpp:
(WebCore::CryptoAlgorithmHKDF::platformDeriveBits):
* Source/WebCore/crypto/openssl/CryptoAlgorithmHMACOpenSSL.cpp:
(WebCore::CryptoAlgorithmHMAC::platformSign):
(WebCore::CryptoAlgorithmHMAC::platformVerify):

Canonical link: https://commits.webkit.org/277888@main
